File: [23.01]Algorithm/baek/src/retry2/B24461.java
package retry2;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class B24461 {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int N = Integer.parseInt(br.readLine());
        int[] cnts = new int[N];
        List<Integer>[] edgeLists = new ArrayList[N];
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < N; i++) {
            edgeLists[i] = new ArrayList<>();
        }

        for (int i = 0; i < N-1; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            edgeLists[a].add(b);
            edgeLists[b].add(a);
            cnts[a]++;
            cnts[b]++;
        }


        System.out.println(sb);

    }

    public static String fn(List<Integer>[] lists, int[] levels, int N) {
        StringBuilder sb = new StringBuilder();
        boolean[] visits = new boolean[N];
        Deque<Integer> dq = search(levels, N);

        while (dq.size() > 2) {
            Deque<Integer> dq2 = new ArrayDeque<>(dq);
            dq.clear();

            while (!dq2.isEmpty()) {
                int cur = dq2.poll();
                visits[cur] = true;

                for (int next :
                        lists[cur]) {
                    if(!visits[next] && --levels[next] == 1){
                        dq.add(next);
                    }
                }
            }
        }

        for (int i = 0; i < N; i++) {
            if (!visits[i]) {
                sb.append(i).append(" ");
            }
        }

        return sb.toString();
    }

    public static Deque<Integer> search(int[] levels, int N) {
        Deque<Integer> dq = new ArrayDeque<>();

        for (int i = 0; i < N; i++) {
            if (levels[i] == 1) {
                dq.offer(i);
            }
        }

        return dq;
    }


}

File: [23.01]Algorithm/kotlin/src/main/kotlin/retry2/B24461.kt
package retry2

fun main() {
    val N = readLine()!!.toInt()
    val lists = Array(N + 1) { ArrayList<Int>() }
    val linkCnt = IntArray(N + 1)

    repeat(N - 1) {
        val (a, b) = readLine()!!.split(" ").map { it.toInt() }
        lists[a].add(b)
        lists[b].add(a)
        linkCnt[a]++
        linkCnt[b]++
    }

    println(fn(N, linkCnt, lists))
}

fun fn(N: Int, linkCnt: IntArray, lists: Array<ArrayList<Int>>): String {
    val dq = search(N, linkCnt)
    val visits = BooleanArray(N + 1)
    val result = mutableListOf<Int>()

    while (dq.size > 2) {
        val dq2 = ArrayDeque(dq)
        dq.clear()

        while (dq2.isNotEmpty()) {
            val cur = dq2.removeLast()
            visits[cur] = true

            for (ints in lists[cur]) {
                if (visits[ints]) {
                    continue
                }

                if (--linkCnt[ints] == 1) {
                    dq.add(ints)
                }
            }
        }
    }

    for (i in 0 until N) {
        if (!visits[i]) {
            result.add(i)
        }
    }

    return result.joinToString(" ")
}

fun search(N: Int, linkCnt: IntArray): ArrayDeque<Int> {
    return ArrayDeque((0 until N).filter { linkCnt[it] == 1 })
}
